adc/read_vsys/power_status.c

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2023 Raspberry Pi (Trading) Ltd.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "stdbool.h"
+#include "hardware/adc.h"
+#include "power_status.h"
+
+#if CYW43_USES_VSYS_PIN
+#include "pico/cyw43_arch.h"
+#endif
+
+#ifndef PICO_POWER_SAMPLE_COUNT
+#define PICO_POWER_SAMPLE_COUNT 3
+#endif
+
+// Pin used for ADC 0
+#define PICO_FIRST_ADC_PIN 26
+
+int power_source(bool *battery_powered) {
+#if defined CYW43_WL_GPIO_VBUS_PIN
+    *battery_powered = !cyw43_arch_gpio_get(CYW43_WL_GPIO_VBUS_PIN);
+    return PICO_OK;
+#elif defined PICO_VBUS_GPIO_PIN
+    gpio_set_function(PICO_VBUS_GPIO_PIN, GPIO_FUNC_SIO);
+    *battery_powered = !gpio_get(PICO_VBUS_GPIO_PIN);
+    return PICO_OK;
+#else
+    return PICO_ERROR_NO_DATA;
+#endif
+}
+
+int power_voltage(float *voltage_result) {
+#ifndef PICO_VSYS_PIN
+    return PICO_ERROR_NO_DATA;
+#endif
+#if CYW43_USES_VSYS_PIN
+    cyw43_thread_enter();
+    // Make sure cyw43 is awake
+    cyw43_arch_gpio_get(CYW43_WL_GPIO_VBUS_PIN);
```

Comment:
`cyw43_arch_gpio_get` communicates with the wifi chip, and also calls to `cyw43_thread_enter` and `cyw43_thread_exit`.

Would that not cause conflicts and or release our lock too soon?

Reply:
Are you seeing an issue? You can nest calls to enter and exit so I wouldn't expect this to be a problem.